crates/engine/tree/src/tree/mod.rs

```diff
@@ -176,31 +176,115 @@ impl TreeState {
         true
     }
 
-    /// Remove all blocks up to __and including__ the given block number.
-    pub(crate) fn remove_before(&mut self, upper_bound: BlockNumber) {
-        let mut numbers_to_remove = Vec::new();
-        for (&number, _) in
-            self.blocks_by_number.range((Bound::Unbounded, Bound::Included(upper_bound)))
-        {
-            numbers_to_remove.push(number);
+    /// Remove single executed block by its hash.
+    ///
+    /// ## Returns
+    ///
+    /// The removed block and the block hashes of its children.
+    fn remove_by_hash(&mut self, hash: B256) -> Option<(ExecutedBlock, HashSet<B256>)> {
+        let executed = self.blocks_by_hash.remove(&hash)?;
+
+        // Remove this block from collection of children of its parent block.
+        let parent_entry = self.parent_to_child.entry(executed.block.parent_hash);
+        if let hash_map::Entry::Occupied(mut entry) = parent_entry {
+            entry.get_mut().remove(&hash);
+
+            if entry.get().is_empty() {
+                entry.remove();
+            }
         }
 
-        for number in numbers_to_remove {
-            if let Some(blocks) = self.blocks_by_number.remove(&number) {
-                for block in blocks {
-                    let block_hash = block.block.hash();
-                    self.blocks_by_hash.remove(&block_hash);
+        // Remove point to children of this block.
+        let children = self.parent_to_child.remove(&hash).unwrap_or_default();
 
-                    if let Some(parent_children) =
-                        self.parent_to_child.get_mut(&block.block.parent_hash)
-                    {
-                        parent_children.remove(&block_hash);
-                        if parent_children.is_empty() {
-                            self.parent_to_child.remove(&block.block.parent_hash);
-                        }
-                    }
+        // Remove this block from `blocks_by_number`.
+        let block_number_entry = self.blocks_by_number.entry(executed.block.number);
+        if let btree_map::Entry::Occupied(mut entry) = block_number_entry {
+            // We have to find the index of the block since it exists in a vec
+            if let Some(index) = entry.get().iter().position(|b| b.block.hash() == hash) {
+                entry.get_mut().swap_remove(index);
```

Comment:
this works, and order doesn't matter, we also only expect 1 entry

alternatively this can be solved with `retain` I think but doesn't matter